linux: support hierarchies and cgroup2 for mem constraint lookup

kjin opened the pull request:
This PR adds support for arbitrary mount locations and hierarchies for cgroups, as well as memory limits from cgroups v2. Fixes #2315

It follows this procedure for resolving `{ cgroups_version, path }` for the `memory` subsystem:

1. Read `/proc/self/mountinfo` to get three fields:
    * the __mount point__ for the `memory` subsystem (usually `/sys/fs/cgroup/memory` for cgroups v1, and `/sys/fs/cgroup` for cgroups v2)
    * the __root__ path of the device being mounted at the location above (usually `/`, but could be `/docker/${CONTAINER_PID}` for docker containers, or any other value)
    * whether the `memory` subsystem is controlled by cgroups v1 or v2 (-> `cgroups_version`)
2. Read `/proc/self/cgroup` to get a "hierarchy path" for the given subsystem. This comes from a line of the form `${ID}:memory:${HIERARCHY}` in cgroups v1 and `0::${HIERARCHY}` in cgroups v2. It _should_ start with the __root__ string mentioned above, and replacing this with the __mount point__ should yield the "true path" to the memory parameter files (-> `path`). For example:
    * If the __root__ found above was `/`, the __mount_point__ was `/sys/fs/cgroup2`, and the hierarchy path read here was `/foo/bar`, then the true path should be `/sys/fs/cgroup2/foo/bar`
    * If the __root__ found above was `/docker/abcdef0`, the __mount_point__ was 
`/sys/fs/cgroup/memory`, and the hierarchy path read here was `/docker/abcdef0/foo/bar`, then the true path should be `/sys/fs/cgroup/memory/foo/bar`

If the attempt to get this true path fails or any assumptions above are violated, we should act as if cgroups is not present and just return `0` to signify no known upper memory limit. Otherwise, we use the information above to read either `memory.limit_in_bytes` (for v1) or `memory.high`/`memory.max` (for v2) from the correct location.

kjin commented:
cc/ @ofrobots, @ChALkeR

ofrobots commented on src/unix/linux-core.c:
```diff
@@ -1014,15 +1015,257 @@ uint64_t uv_get_total_memory(void) {
+  const char* field_start[10];
+  size_t field_len[10];
+  size_t field_idx;
+  char root[1024]; /* (4) */
```
It would be good to be consistent with the buffers used for paths. For example, why is this short than `info->path`.

ofrobots commented on src/unix/linux-core.c:
```diff
@@ -1014,15 +1015,257 @@ uint64_t uv_get_total_memory(void) {
+  /* From /proc/self/cgroup */
+  char hierarchy_path[1024];
+  /* Should be big enough to fit a subsystem name, plus 2 more characters. */
+  char subsystem_search_string[64];
```
Is it documented somewhere that subsystems will be no more than 62 bytes?

kjin replied:
It should actually be no more than 10, the longest name is "perf_event". I don't think there are arbitrary subsystems (at least any with parameter files) so that's as long as it should get. I've reduced this down to `16`.

ofrobots replied:
Clean code is obviously correct. Instead of relying on implementation details, I think you should write this in a way that doesn't depending on deep knowledge and point-in-time implementation details. Code should be obviously correct to a casual reader.

My suggestion is to allocate this dynamically based on the length of the subsystem string passed in instead of relying on constants.

kjin commented on src/unix/linux-core.c:
```diff
@@ -1014,15 +1015,257 @@ uint64_t uv_get_total_memory(void) {
+#define CGROUPS_VERSION_1 0x1
+#define CGROUPS_VERSION_2 0x2
+#define CGROUPS_V2_CTRL_PREFIX "0::"
+#define CGROUPS_V2_CTRL_PREFIX_LEN strlen(CGROUPS_V2_CTRL_PREFIX)
```
I've made it `3`, this value should not change over time anyway.

kjin commented on src/unix/linux-core.c:
```diff
@@ -1014,15 +1015,257 @@ uint64_t uv_get_total_memory(void) {
+  /* cgroups version, one of CGROUPS_VERSION_*. */
+  uint8_t cgroups_version;
+  /* Path in which param files are found for a subsystem. */
+  char path[4096];
```
Ok, I've made this a macro, and any field that could potentially contain a path is now 4096 bytes long.

ofrobots commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1016,261 @@ uint64_t uv_get_total_memory(void) {
+ * remainder of the string (excluding the whitespace).
+ * Note that we assume that `start` ends with "\n\0".
+ */
+static const char* uv__mountinfo_next_field(const char* start,
```
I would suggest replacing this function with `strsep`. Since that is not portable, you can implement your own version, or use a public domain copy.  It would be better to have a copy of `strsep` rathan than a function that is functionally equivalent but subtly different.

ofrobots commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1016,261 @@ uint64_t uv_get_total_memory(void) {
+ * `super_options_len`), return TRUE if option exactly matches one of the
+ * elements in that sequence.
+ */
+static int uv__mountinfo_find_super_option(const char* super_options,
```
Once you have `strsep`, this would be easy to write in terms of that.

ChALkeR commented:
This looks great at the first glance, thanks!
I will be able to look closer and test this out tomorrow =).

ofrobots commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1016,261 @@ uint64_t uv_get_total_memory(void) {
+   * Henceforth, numbers in parentheses refer to fields from the docs.
+   */
+  const char* field_ptr;
+  const char* field_start[10];
```
Use named fields instead. Since you will be using strsep, they would be null terminated, so you don't need to have a `field_len` array.

ofrobots commented:
I think this code can be rewritten in terms of `strsep` and the central loop parsing individual lines in `mountinfo` will simplified / easier to read once it uses named fields.

I'll re-review once those changes are done.

kjin commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1016,261 @@ uint64_t uv_get_total_memory(void) {
+  char hierarchy_path[CGROUPS_BUF_SIZE];
+  /*
+   * Should be big enough to fit a subsystem name, plus 2 more characters.
+   * The longest known subsystem name (from cgroups docs) is "perf_event".
```
Made this dynamically alloc'd.

ChALkeR commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1016,217 @@ uint64_t uv_get_total_memory(void) {
+    return 0;
+  }
+  hierarchy_path_inner = hierarchy_path + strlen(root);
+  snprintf(info->path, CGROUPS_BUF_SIZE, "%s/%s", mount_point, hierarchy_path_inner);
```
I get a gcc `format-truncation` warning here on gcc v8.3.0.

ChALkeR commented:
On two subsequent calls to `uv_get_constrained_memory`, this is what I observe in `info.path`:
1) Length = `103`, value = `/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/run-r0858a2015da94808831f65532cf9d476.scope`.
2) Length = `1233`, value = 
```
/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/run-r0858a2015da94808831f65532cf9d476.scopeon):     840500 kB
Inactive(anon):    50100 kB
Active(file):     814984 kB
Inactive(file):   620792 kB
Unevictable:      658560 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:               112 kB
Writeback:             0 kB
AnonPages:        839480 kB
Mapped:           537664 kB
Shmem:            709808 kB
KReclaimable:     148448 kB
Slab:             208096 kB
SReclaimable:     148448 kB
SUnreclaim:        59648 kB
KernelStack:        6308 kB
PageTables:        19476 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     8164272 kB
Committed_AS:    4379052 kB
VmallocTotal:   34359738367 kB
VmallocUsed:           0 kB
VmallocChunk:          0 kB
Percpu:             1504 kB
HardwareCorrupted:     0 kB
AnonHugePages:         0 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
DirectMap4k:      131472 kB
DirectMap2M:     6064128 kB
DirectMap1G:    10485760 kB

```

ChALkeR commented on src/unix/linux-core.c:
```diff
@@ -1045,10 +1252,27 @@ static uint64_t uv__read_cgroups_uint64(const char* cgroup, const char* param) {
+    high = uv__read_cgroups_uint64(info.path, "memory.high");
+    if (max == 0 || high == 0)
+      return 0;
+    return max > high ? max : high;
```
Why does this return the maximum of the two? Shouldn't it return the minumum?

kjin replied:
Yes, you're right.

ChALkeR commented on src/unix/linux-core.c:
```diff
@@ -1045,10 +1252,27 @@ static uint64_t uv__read_cgroups_uint64(const char* cgroup, const char* param) {
+    max = uv__read_cgroups_uint64(info.path, "memory.max");
+    high = uv__read_cgroups_uint64(info.path, "memory.high");
+    if (max == 0 || high == 0)
+      return 0;
```
Why does this return 0 when _any_ of those are 0? That effectively breaks the testcase -- i.e. if only `MemoryHigh` or only `MemoryMax` is set, this function returns 0.

kjin replied:
Thanks, I fixed it so that we only get zero if they're both zero, and if there is one non-zero value, the other is used.

ChALkeR commented:
Overall, this seems to work after those issues would be fixed.

ofrobots commented:
@kjin is on vacation this week + next. This will be picked back up once he's back.

kjin commented:
Re: `info.path` Corruption -- I am still trying to figure out why this is the case. I can't seem to figure out how to run with `systemd-run --user`, at least over ssh. Will look into it more over the next day.

ChALkeR commented:
@kjin which system are you using and did you specify `systemd.unified_cgroup_hierarchy=1` kernel option?

kjin commented:
@ChALkeR I'm using Debian stretch. did specify `systemd.unified_cgroup_hierarchy=1` and was able to see the v2 hierarchy. But I wasn't able to reproduce the bug that you mentioned, and I assume it was because I couldn't pass `--user` to `systemd-run` (it seems to work, even on consecutive calls, without this arg).

kjin commented:
Ok, I was able to get it to reproduce on a local Linux machine.

kjin commented:
I've added a commit that handles the case where a line in `/proc/self/[mountinfo|cgroup]` is too long (we skip it; the max line length is also now set to 16KB, which I believe is a hard max for lines with information relevant to cgroups).

ofrobots commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1016,281 @@ uint64_t uv_get_total_memory(void) {
+  strcpy(haystack_mutable, haystack);
+  haystack_ptr = haystack_mutable;
+  do {
+    candidate = strsep(&haystack_ptr, &separator);
```
AFAICT `strsep` expects the delimiter to be a null terminated string rather than the address of a char.

ofrobots commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1016,281 @@ uint64_t uv_get_total_memory(void) {
+ * Returns 0 if the entire line is read into `buf`, a positive number if it was
+ * truncated, and UV_EIO if reading the file using fgets returns NULL.
+ */
+static int uv__read_line(char* buf, size_t buf_size, FILE* fp) {
```
IMO this would be easier to implement in terms of `getline`. This whole function might become unnecessary..

kjin replied:
Done, but the current commit has a bug in it which I'm aiming to fix by EOD tomorrow.

kjin replied:
Ok, it's been fixed.

ofrobots commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1016,281 @@ uint64_t uv_get_total_memory(void) {
+   * See description of /proc/[pid]/mountinfo in proc(5).
+   * Henceforth, numbers in parentheses refer to fields from the docs.
+   */
+  char* field_ptr;
```
This is only used inside the while loop below. Move the declaration there.

kjin commented:
Hi @bnoordhuis @cjihrig (no specific reason other than that you're people that I think I've interacted in the past on this repo), I think this change is ready for a broader review. Thanks!

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1034,7 +1253,10 @@ static uint64_t uv__read_cgroups_uint64(const char* cgroup, const char* param) {
     buf[n] = '\0';
-    sscanf(buf, "%" PRIu64, &rc);
+    if (!strcmp(buf, CGROUPS_V2_NO_LIMIT_VALUE))
+      rc = 0;
```
`rc` is initialized to 0 a few lines up, so this `if-else` can probably be rewritten as just an `if (0 != strcmp(buf, CGROUPS_V2_NO_LIMIT_VALUE))`

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1045,10 +1267,40 @@ static uint64_t uv__read_cgroups_uint64(const char* cgroup, const char* param) {
+  uv__cgroups_subsystem_info_t info;
+  uint64_t rc;
+  /* For v2 only. */
+  uint64_t max, high;
```
Tiny nit, but can you make this two separate variables.

kjin commented:
@cjihrig Thanks for looking! I've addressed your comments in the most recent commit.

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+  /*
+   * Loop once per line; try to find the mount location for the given subsystem.
+   */
+  while (!feof(fp)) {
```
`feof(fp) == 0`

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+     */
+    if (strcmp(curr_fs_type, "cgroup") == 0) {
+      /* cgroups v1 */
+      if (uv__find_in_delimited_string(curr_super_options, subsystem, ",")) {
```
`uv__find_in_delimited_string(curr_super_options, subsystem, ",") == 1`

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+  subsystem_search_string = malloc(strlen(subsystem) + 3);
+  snprintf(subsystem_search_string, strlen(subsystem) + 3, ":%s:", subsystem);
+
+  while (!feof(fp)) {
```
`feof(fp) == 0`

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+
+    if (CGROUPS_VERSION_1 == info->cgroups_version) {
+      hierarchy_path_search_ptr = strstr(buf, subsystem_search_string);
+      if (hierarchy_path_search_ptr) {
```
No need for the curly braces here.

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+      if (strncmp(buf, CGROUPS_V2_CTRL_PREFIX, CGROUPS_V2_CTRL_PREFIX_LEN) == 0)
+        hierarchy_path_search_ptr = buf + CGROUPS_V2_CTRL_PREFIX_LEN;
+    }
+    if (hierarchy_path_search_ptr) {
```
`hierarchy_path_search_ptr  != NULL`

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+   * and should be replaced with the mount point.
+   */
+  if (strncmp(hierarchy_path, root, strlen(root)) == 0) {
+    const char* hierarchy_path_inner;
```
Can you hoist these variables to the top of the function.

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+  }
+
+  fclose(fp);
+  if (NULL != buf)
```
No need for this `if`. `free(NULL)` is fine.

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+
+  fclose(fp);
+  free(subsystem_search_string);
+  if (NULL != buf)
```
No need for the `if` here either.

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+
+file_malformed:
+  fclose(fp);
+  if (NULL != buf)
```
You can drop the `if` here as well.

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,231 @@ uint64_t uv_get_total_memory(void) {
+  if (NULL != buf)
+    free(buf);
+  info->cgroups_version = CGROUPS_VERSION_UNKNOWN;
+  if (subsystem_search_string != NULL)
```
No need for the `if` here.

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1045,10 +1273,37 @@ static uint64_t uv__read_cgroups_uint64(const char* cgroup, const char* param) {
+
+  rc = 0;
+
+  if (!uv__read_cgroups_proc_files(&info, "memory")) {
```
`uv__read_cgroups_proc_files(&info, "memory") == 0`

cjihrig commented on src/unix/linux-core.c:
```diff
@@ -1045,10 +1273,37 @@ static uint64_t uv__read_cgroups_uint64(const char* cgroup, const char* param) {
+        rc = max < high ? max : high;
+    }
+
+    if (NULL != info.path)
```
No need for the `if`.

cjihrig commented:
cc: @libuv/collaborators

cjihrig commented:
@kjin it looks like there are relevant CI failures.

saghul commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,226 @@ uint64_t uv_get_total_memory(void) {
+  haystack_size = strlen(haystack) + 1;
+  if (needle == NULL || strlen(needle) >= haystack_size)
+    return 0;
+  haystack_mutable = malloc(haystack_size);
```
use `uv__malloc` / `uv__free`

saghul commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,226 @@ uint64_t uv_get_total_memory(void) {
+  size_t buf_size;
+
+  /* From /proc/self/mountinfo */
+  char root[PATH_MAX];
```
use `UV__PATH_MAX`

saghul commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,226 @@ uint64_t uv_get_total_memory(void) {
+
+
+/*
+ * Read /proc/self/mountinfo and /proc/self/cgroup to get info about how the
```
putthing some sample output of those 2 files here as documentation would be helpful

saghul commented on src/unix/linux-core.c:
```diff
@@ -1045,10 +1268,36 @@ static uint64_t uv__read_cgroups_uint64(const char* cgroup, const char* param) {
+  uv__cgroups_subsystem_info_t info;
+  uint64_t rc;
+  /* For v2 only. */
+  uint64_t max;
```
maybe declare them in the if block?

kjin replied:
@cjihrig recommended hoisting var declarations to the top of a function from within a while loop, does that also apply to if blocks?

vtjnash replied:
either is acceptable, but top of the function seems to be usually preferred in the style guide

cjihrig commented:
@saghul are you OK with this now?

saghul commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,263 @@ uint64_t uv_get_total_memory(void) {
+  /* cgroups version, one of CGROUPS_VERSION_*. */
+  uint8_t cgroups_version;
+  /* Path in which param files are found for a subsystem. */
+  char* path;
```
style (here ald elsewhere): `char *path`

cjihrig replied:
Hasn't the libuv style always been "the star leans left?"

saghul replied:
Yeah sorry, brainfart.

saghul commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,263 @@ uint64_t uv_get_total_memory(void) {
+  haystack_size = strlen(haystack) + 1;
+  if (needle == NULL || strlen(needle) >= haystack_size)
+    return 0;
+  haystack_mutable = uv__malloc(haystack_size);
```
maybe this can be replaced with `uv__strndup` ? https://github.com/libuv/libuv/blob/89a027d862c9b84a20b0316415be169ac25dbf01/src/uv-common.c#L63

kjin replied:
Done

saghul commented:
LGTM. Left a minor suggestion, not a blocker. Great work!

cjihrig commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1484/

It looks like there are some failures. Just a heads up, @kjin - there is a release scheduled for this week in case you want to try to get this in the next release.

kjin commented:
@saghul @cjihrig Thanks for the look!

It seems like I forgot to rebase which accounted for the Unix failures (`UV__PATH_MAX` was introduced in-between). It should be working now. I do get a 404 when viewing the Windows CI though, so I don't know what's going on there.

saghul commented:
New CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1485/

bnoordhuis commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,262 @@ uint64_t uv_get_total_memory(void) {
+    return UV__ERR(errno);
+  
+  /* + 3 for two colons, and terminal character. */
+  subsystem_search_string = uv__malloc(strlen(subsystem) + 3);
```
Check for NULL.

bnoordhuis commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,262 @@ uint64_t uv_get_total_memory(void) {
+      goto file_malformed;
+    }
+    if ('\n' == buf[strlen(buf) - 1])
+      buf[strlen(buf) - 1] = '\0';
```
If you cache the result of `getline()`, you don't need to call `strlen()` twice. (Also applies to line 1155-1156.)

bnoordhuis commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,262 @@ uint64_t uv_get_total_memory(void) {
+
+    if (CGROUPS_VERSION_1 == info->cgroups_version) {
+      hierarchy_path_search_ptr = strstr(buf, subsystem_search_string);
+      if (hierarchy_path_search_ptr)
```
`hierarchy_path_search_ptr != NULL`

bnoordhuis commented on src/unix/linux-core.c:
```diff
@@ -75,6 +75,16 @@
+#define CGROUPS_VERSION_1 0x1
+#define CGROUPS_VERSION_2 0x2
+#define CGROUPS_V2_CTRL_PREFIX "0::"
+#define CGROUPS_V2_CTRL_PREFIX_LEN 3
```
Prefer `(sizeof(CGROUPS_V2_CTRL_PREFIX) - 1)`

Aside: these defines only seem to be used in one place so I'd be okay with open-coding them. That way a reader of the code doesn't have to look them up.

bnoordhuis commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,262 @@ uint64_t uv_get_total_memory(void) {
+    hierarchy_path_inner = hierarchy_path + strlen(root);
+    /* +2 for "/" and null terminator. */
+    path_size = strlen(mount_point) + strlen(hierarchy_path_inner) + 2;
+    info->path = uv__malloc(path_size);
```
Check for NULL.

bnoordhuis reviewed (dismissed):
The string manipulation and memory handling still needs some work, I left some comments.

Using the red button to counteract the LGTMs.

kjin commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,262 @@ uint64_t uv_get_total_memory(void) {
+ * Given a `separator`-delimited string `haystack`, return 1 if `needle` exactly
+ * matches at least one of the elements in that sequence.
+ */
+static int uv__find_in_delimited_string(const char* haystack, const char* needle, const char* separator) {
```
I think I still have comments (that represent lines in `/proc/self/mountinfo`) that are more than 80 columns long, should I wrap those as well?

vtjnash replied:
Yes

vtjnash replied:
(unless there's a specific reason for exemption, such as being a literal copy of another file)

kjin replied:
It's terminal output, so I wasn't sure -- that's a bit like being a literal copy. Referring specifically to [here](https://github.com/libuv/libuv/pull/2323/files#diff-cf4c0dad7cbbf7751061ba552c3c50ef46abdeaaca477a491845d7e254f57b92R790-R792):

Should I still wrap these?

kjin commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,262 @@ uint64_t uv_get_total_memory(void) {
+  }
+
+  fclose(fp);
+  uv__free(buf);
```
Re the aside: I assume `free` would be more suitable then, let me know if this is not the case.

kjin commented on src/unix/linux-core.c:
```diff
@@ -1015,14 +1025,262 @@ uint64_t uv_get_total_memory(void) {
+  haystack_mutable = uv__strndup(haystack, haystack_len);
+  haystack_ptr = haystack_mutable;
+  do {
+    candidate = strsep(&haystack_ptr, separator);
```
According to `strsep` docs, its return value is the value pointed to in the first argument, or in this case `haystack_ptr`. Would the null check still be necessary if we are (now) ensuring elsewhere that the input, `haystack_ptr`, is not null?

vtjnash replied:
I will say not needed

kjin commented:
@bnoordhuis I've updated the PR making the requested changes. In particular, since it was a relatively large change, the change to malloc large strings is in its separate commit.

kjin commented:
Gentle ping @bnoordhuis, whenever you get a chance to look at this it would be much appreciated!

cjihrig commented:
@bnoordhuis can you take another look at this one?

BridgeAR commented:
@bnoordhuis PTAL. This is blocking https://github.com/nodejs/node/pull/27805

ChALkeR commented:
Ecosystem status update:
* Fedora 31 [was released in October](https://fedoramagazine.org/announcing-fedora-31/), migrated to `unified` hierarchy by default
* Mandriva is also migrating to `unified` hierarchy by default.

Without this PR, `uv_get_constrained_memory` is broken on those systems.

vtjnash commented:
Ping @bnoordhuis: is this done now (other than needing a rebase to fix a trivial merge conflict)

maleadt commented:
Seeing how this PR was pretty much good to go, could somebody else review the latest changes? @vtjnash maybe?
@kjin Would you be willing to rebase the PR so that we can move forward with it?

kjin commented:
Let me give it a try!

kjin commented:
Ok, I've rebased and resolved the conflict -- which occurred due to part of `uv__read_cgroups_uint64` being refactored into a common `uv__slurp` function in 1a890036351897967a77849f4709a98b70c461ef. If CI passes I believe this should be ready to go!

kjin commented:
Hi @bnoordhuis, do you mind reviewing this PR, or at least resolving the change request? It seems like quite a few people have been hoping to seeing this land. If you're busy, it would be good to perhaps assign a different reviewer to help move this forward? Thanks!

maleadt commented:
@vtjnash maybe?

kjin commented:
If this PR remains stuck on the change-requested flag even after a third reviewer, I'd be inclined to open a new one, but I'd like to get blessing for that by an owner of libuv first.

vtjnash commented on src/unix/linux-core.c:
```diff
@@ -725,27 +732,383 @@ uint64_t uv_get_total_memory(void) {
+      continue;
+    strsep(&field_ptr, " "); /* mount options */
+    /* A hyphen marks the end of variable-length optional fields. */
+    while ('-' != field_ptr[0])
```
```suggestion
    while (field_ptr != NULL && '-' != field_ptr[0])
```

vtjnash commented on src/unix/linux-core.c:
```diff
@@ -725,27 +732,383 @@ uint64_t uv_get_total_memory(void) {
+
+  char filename[UV__PATH_MAX];
   char buf[32];  /* Large enough to hold an encoded uint64_t. */
   uint64_t rc;
```
`rc` is now undefined on some code paths
```suggestion
  uint64_t rc = 0;
```

vtjnash commented:
Most of @bnoordhuis comments were addressed, but a few seem to have been missed. I'll dismiss his review (and substitute my own approval), once my comments are addressed.

kjin commented:
@vtjnash Thank you for the review -- I haven't forgotten about this, but unfortunately won't be able to circle back until next week.

kjin commented:
@vtjnash OK, I've addressed your comments, both the news ones and the ones you left on Ben's review. I think there is one open question I had regarding wrapping terminal output when it's in a comment ([context](https://github.com/libuv/libuv/pull/2323#discussion_r722721185)).

bnoordhuis commented:
@kjin Thanks for picking this up again. Cgroups v2 support is valuable IMO but I'd like to cut down the amount of parsing libuv has to do; C is not a great language for that.

You parse `/proc/self/mountinfo` to determine if the system uses cgroups v1 or v2? Is that necessary? Doesn't the format of the `memory` field in `/proc/self/cgroups` tell you that?

Would `open("/sys/fs/cgroup/unified", O_DIRECTORY|O_CLOEXEC)` work? It's a v2 system when it's there, it seems reasonable to assume v1 when it's not. (Related subject: I prefer openat() over path munging.)

Let me know if you don't have the time or inclination to return to this and I'll try to adopt it.

kjin commented:
Hi @bnoordhuis, thanks for your input. I'll have bandwidth in around ~2 weeks (after American Thanksgiving) to work on this. If you (or anyone else) feel inclined to move this forward before then, please go ahead -- I want to make sure I'm not blocking anything critical. But I haven't forgotten about this PR, and I'll post back here again when I'm starting to work on it.

maleadt commented:
Since this has been marked as stale, and I don't have the necessary permissions to push here, I've opened a new pull request where I've rebased these changes and addressed one of the remaining review comments: https://github.com/libuv/libuv/pull/3712

kjin commented:
@maleadt Thank you for moving this forward. A lot of changes have happened since I opened this PR that have brought me far away from the Node.js ecosystem. I'm sorry that I couldn't address this in a more timely manner, but glad to see that you volunteered to take it over!
